agent/grpc-external/services/peerstream/stream_resources.go

```diff
@@ -251,26 +257,21 @@ func (s *Server) realHandleStream(streamReq HandleStreamRequest) error {
 		}
 	}
 
-	// TODO(peering): Should this be buffered?
-	recvChan := make(chan *pbpeerstream.ReplicationMessage)
+	recvCh := make(chan *pbpeerstream.ReplicationMessage)
+	errCh := make(chan error)
 	go func() {
-		defer close(recvChan)
 		for {
 			msg, err := streamReq.Stream.Recv()
-			if err == nil {
-				logTraceRecv(logger, msg)
-				recvChan <- msg
-				continue
+			if err != nil {
+				errCh <- err
+				return
 			}
-
-			if err == io.EOF {
-				logger.Info("stream ended by peer")
-				status.TrackRecvError(err.Error())
+			logTraceRecv(logger, msg)
+			select {
+			case recvCh <- msg:
+			case <-ctx.Done():
```

Comment:
Should the heartbeat goroutine below also use this `ctx`? Currently it uses `streamReq.Stream.Context()` but this `ctx` will kill the loop faster.

Reply:
Yep, added!